[aot autograd] collect static parameter metadata when graphs fallback to inference

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129116
* #129181
* #128987
* #128982
* __->__ #128905
* #127960

https://github.com/pytorch/pytorch/pull/126820 but for graphs that have requires_grad inputs but no requires_grad outputs i.e. inference graph

the implementation of inference graph fallback was throwing away the static parameter information during metadata recomputation

also adding a cudagraphs counter to test this easier


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

xmfan committed:
[aot autograd] collect static parameter metadata when graphs fallback to inference

[ghstack-poisoned]

xmfan commented on torch/_functorch/aot_autograd.py:
```diff
@@ -602,24 +602,23 @@ def convert(idx, x):
-                            is_train=needs_autograd,
+                            is_train=False,
                             tokens=fw_metadata.tokens,
+                            static_parameter_indices=fw_metadata.static_parameter_indices,
```
this adds the static params into the reconstructed metadata. above lines are just logic simplifications since line 596 sets needs_autograd to False

xmfan committed:
Update on "[aot autograd] collect static parameter metadata when graphs fallback to inference"


https://github.com/pytorch/pytorch/pull/126820 but for graphs that have requires_grad inputs but no requires_grad outputs i.e. inference graph

the implementation of inference graph fallback was throwing away the static parameter information during metadata recomputation

also adding a cudagraphs counter to test this easier


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[aot autograd] collect static parameter metadata when graphs fallback to inference"


https://github.com/pytorch/pytorch/pull/126820 but for graphs that have requires_grad inputs but no requires_grad outputs i.e. inference graph

the implementation of inference graph fallback was throwing away the static parameter information during metadata recomputation

also adding a cudagraphs counter to test this easier


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -955,6 +955,9 @@ def _copy_inputs_and_remove_from_src(self, dsts, srcs):
         # Fails on empty lists
         if dst_tensors:
             torch._foreach_copy_(dst_tensors, src_tensors)
+            counters["inductor"]["cudagraph_copies_due_to_non_static_inputs"] += len(
+                dst_tensors
+            )
 
```
We should not be updating counters on the hot path. can you set in this in CudaGraphNode `__init__` ?

xmfan replied:
moving this into __init__ needs to change the semantics to "cudagraph_recorded_non_static_inputs"

eellison replied:
Just use len(`non_static_input_idx`)

xmfan replied:
yeah i'll use that, it just won't be as informative, since recordings can be thrown away, and it no longer measures how many times the copies happen

eellison commented on torch/_functorch/aot_autograd.py:
```diff
@@ -602,24 +602,23 @@ def convert(idx, x):
-                            keep_input_mutations=aot_config.keep_inference_input_mutations
-                            and not needs_autograd,
-                            is_train=needs_autograd,
+                            keep_input_mutations=aot_config.keep_inference_input_mutations,
```
cc @bdhirsh i dont know this part well enough

bdhirsh replied:
seems ok

xmfan committed:
Update on "[aot autograd] collect static parameter metadata when graphs fallback to inference"


https://github.com/pytorch/pytorch/pull/126820 but for graphs that have requires_grad inputs but no requires_grad outputs i.e. inference graph

the implementation of inference graph fallback was throwing away the static parameter information during metadata recomputation

also adding a cudagraphs counter to test this easier


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[aot autograd] collect static parameter metadata when graphs fallback to inference"


https://github.com/pytorch/pytorch/pull/126820 but for graphs that have requires_grad inputs but no requires_grad outputs i.e. inference graph

the implementation of inference graph fallback was throwing away the static parameter information during metadata recomputation

also adding a cudagraphs counter to test this easier


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
